### PR TITLE
Add documentation for materialized views

### PR DESCRIPTION
**Pull request by mosabua**
Related to the SQL support docs in the Iceberg connector. 

As a follow up after this I will update links and add more info in iceberg connector

### CODE REVIEW

File: docs/src/main/sphinx/sql/create-materialized-view.rst [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+clarifies the content::
+
+    CREATE MATERIALIZED VIEW cancelled_orders
+    AS
+        SELECT orderkey, totalprice
+        FROM orders
+        WHERE orderstatus = 3
+
+Create or replace a materialized view ``order_totals_by_date`` that summarizes
+``orders`` across all orders from all customers. The view creation is
```
Comment (raunaqmorarka) [inline docs/src/main/sphinx/sql/create-materialized-view.rst:61]: We should clarify that CREATE MV AS doesn't actually run the query and store the results. It just creates the MV definition. REFRESH MV must be run to populate the storage table with the query results.

File: docs/src/main/sphinx/sql/create-materialized-view.rst [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+Description
+-----------
+
+Create a new materialized view of a :doc:`select` query. The view is a physical
+manifestation of the query results at time of creation. The data is stored and
+can be referenced by future queries. Since the data is stored directly, these
+queries are typically faster than retrieving data from a view created with the
+same query.
+
+However the data in the materialized becomes out of sync with the underlying
```
Comment (raunaqmorarka) [inline docs/src/main/sphinx/sql/create-materialized-view.rst:25]: When the data in the source tables referenced in the materialized view query changes, querying the materialized view may either return stale results or the results from its execution as a logical view. This depends on the specific implementation of the connector supporting materialized views.

File: docs/src/main/sphinx/sql/create-materialized-view.rst [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    CREATE [ OR REPLACE ] MATERIALIZED VIEW
+    [ IF NOT EXISTS ] view_name
+    [ COMMENT string ]
+    [ WITH properties ]
+    AS query
+
+Description
+-----------
+
+Create a new materialized view of a :doc:`select` query. The view is a physical
```
Comment (jhlodin) [inline docs/src/main/sphinx/sql/create-materialized-view.rst:19]: This view is a...

File: docs/src/main/sphinx/sql/create-materialized-view.rst [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+manifestation of the query results at time of creation. The data is stored and
+can be referenced by future queries. Since the data is stored directly, these
+queries are typically faster than retrieving data from a view created with the
+same query.
+
+However the data in the materialized becomes out of sync with the underlying
+data in the source tables, when that data changes. The data in the materialized
```
Comment (jhlodin) [inline docs/src/main/sphinx/sql/create-materialized-view.rst:26]: Pending your approach to Raunaq's comment, would rephrase as "When this data changes, the materialized view becomes out of sync with the source tables." Circumstance -> Outcome and not the other way around

File: docs/src/main/sphinx/sql/create-materialized-view.rst [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+view is updated with the :doc:`refresh-materialized-view` statement.
```
Comment (jhlodin) [inline docs/src/main/sphinx/sql/create-materialized-view.rst:27]: Refresh the data in the materialized view with the...

File: docs/src/main/sphinx/sql/create-materialized-view.rst [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+The optional ``OR REPLACE`` clause causes the materialized view to be replaced
+if it already exists rather than raising an error.
+
+The optional ``IF NOT EXISTS`` clause checks for availability of a materialized
+view of the same name and skips creation and execution of the query if a
+materialized view exists.
```
Comment (jhlodin) [inline docs/src/main/sphinx/sql/create-materialized-view.rst:34]: Feels redundant, would just say "if one exists" or similar

File: docs/src/main/sphinx/sql/create-materialized-view.rst [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+same query.
+
+However the data in the materialized becomes out of sync with the underlying
+data in the source tables, when that data changes. The data in the materialized
+view is updated with the :doc:`refresh-materialized-view` statement.
+
+The optional ``OR REPLACE`` clause causes the materialized view to be replaced
+if it already exists rather than raising an error.
+
+The optional ``IF NOT EXISTS`` clause checks for availability of a materialized
```
Comment (jhlodin) [inline docs/src/main/sphinx/sql/create-materialized-view.rst:32]: Would prefer "checks if a materialized view... exists". Current wording feels roundabout.

File: docs/src/main/sphinx/sql/create-materialized-view.rst [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+view of the same name and skips creation and execution of the query if a
+materialized view exists.
+
+The optional ``COMMENT`` clause causes a ``string`` comment to be stored with
+the metadata about the materialized view. It can be displayed with the
```
Comment (jhlodin) [inline docs/src/main/sphinx/sql/create-materialized-view.rst:37]: "It can be displayed" -> "This comment is displayed"

File: docs/src/main/sphinx/sql/create-materialized-view.rst [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+:doc:`show-create-materialized-view` statement and is available in
+``system.metadata.materialized_view_properties``.
+
+The optional ``WITH`` clause is used define properties to use for the
```
Comment (jhlodin) [inline docs/src/main/sphinx/sql/create-materialized-view.rst:41]: ``WITH`` clause defines properties

File: docs/src/main/sphinx/sql/create-materialized-view.rst [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+materialized view creation. Separate multiple property/value pairs by commas.
+The supported properties are different for each connector and detailed in the
+SQL support section of the specific connector documentation.
+
+Examples
+--------
+
+Create a simple materialized view ``cancelled_orders`` over the ``orders`` table
```
Comment (jhlodin) [inline docs/src/main/sphinx/sql/create-materialized-view.rst:49]: Personal preference: The following statement creates

File: docs/src/main/sphinx/sql/create-materialized-view.rst [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+    CREATE MATERIALIZED VIEW cancelled_orders
+    AS
+        SELECT orderkey, totalprice
+        FROM orders
+        WHERE orderstatus = 3
+
+Create or replace a materialized view ``order_totals_by_date`` that summarizes
+``orders`` across all orders from all customers. The view creation is
+potentially processing intensive, while reading from the materialized view
+subsequently is a simple read operation::
```
Comment (jhlodin) [inline docs/src/main/sphinx/sql/create-materialized-view.rst:63]: Feels like this could be more direct about what it's saying. Consider:

"Note that the view creation can be processing intensive, but reading from the materialized view is a simple read operation::"

File: docs/src/main/sphinx/sql/drop-materialized-view.rst [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+.. code-block:: text
+
+    DROP MATERIALIZED VIEW [ IF EXISTS ] view_name;
+
+Description
+-----------
+
+Drop an existing materialized view.
+
+The optional ``IF EXISTS`` clause causes the error to be suppressed if
```
Comment (jhlodin) [inline docs/src/main/sphinx/sql/drop-materialized-view.rst:17]: Would love a specific error message or code here if possible, rather than just "the error". Better for SEO.

File: docs/src/main/sphinx/sql/refresh-materialized-view.rst [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+Description
+-----------
+
+Refresh the data stored in the materialized view ``view_name``. Data is
+retrieved from the underlying tables accessed by the query that creates the
+materialized view.
+
+The specific implementation varies by connector used to create the materialized
+view. The simplest implementation drops the materialized view and creates it
+again. More complex implementation insert missing data and update changed data.
```
Comment (jhlodin) [inline docs/src/main/sphinx/sql/refresh-materialized-view.rst:21]: I feel like talking about "complexity" adds an unnecessary dimension to this concept. Would rephrase as

"Some connectors drop the materialized view and create it again, while others insert missing data and update changed data."

**Review by jhlodin — CHANGES_REQUESTED** [review]

File: docs/src/main/sphinx/sql/create-materialized-view.rst (hunk 1 above)
Comment (mosabua) [reply docs/src/main/sphinx/sql/create-materialized-view.rst:61]: Oh .. I was not aware of that.

File: docs/src/main/sphinx/sql/create-materialized-view.rst (hunk 2 above)
Comment (mosabua) [reply docs/src/main/sphinx/sql/create-materialized-view.rst:25]: What? Are you saying that a parallel logical view is created?

Comment (raunaqmorarka) [reply docs/src/main/sphinx/sql/create-materialized-view.rst:25]: The MV behaves just like a VIEW when the connector tells the engine that it is stale. So a query referencing the MV will just end up running the whole query as if it was referencing a VIEW.

File: docs/src/main/sphinx/sql/refresh-materialized-view.rst (hunk 13 above)
Comment (raunaqmorarka) [reply docs/src/main/sphinx/sql/refresh-materialized-view.rst:21]: Currently the only implementation in Trino is Iceberg and that simply drops the existing table and creates a new one with the results of executing the MV query.

File: docs/src/main/sphinx/sql/create-materialized-view.rst (hunk 3 above)
Comment (anjalinorwood) [inline docs/src/main/sphinx/sql/create-materialized-view.rst:19]: This materialized view is a physical ..

File: docs/src/main/sphinx/sql/create-materialized-view.rst [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+
+    CREATE MATERIALIZED VIEW cancelled_orders
+    AS
+        SELECT orderkey, totalprice
+        FROM orders
+        WHERE orderstatus = 3
+
+Create or replace a materialized view ``order_totals_by_date`` that summarizes
+``orders`` across all orders from all customers. The view creation is
+potentially processing intensive, while reading from the materialized view
```
Comment (anjalinorwood) [inline docs/src/main/sphinx/sql/create-materialized-view.rst:62]: Materialized view creation is a metadata operation and is NOT processing intensive. The 'refresh materialized view' command is potentially processing intensive. Reading from the materialized view is a simple read operation.

File: docs/src/main/sphinx/sql/create-materialized-view.rst [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+        SELECT orderdate, sum(totalprice) AS price
+        FROM orders
+        GROUP BY orderdate
+
+Create a materialized view with a comment, partitioning in the storage and a
+complex query using ``WITH``::
+
+    CREATE MATERIALIZED VIEW example_mat_view
+    COMMENT 'A partitioned materialized view'
+    WITH ( partitioned_by = ARRAY ['dateint'] )
```
Comment (anjalinorwood) [inline docs/src/main/sphinx/sql/create-materialized-view.rst:76]: Replace 'partitioned_by' by 'partitioning'

File: docs/src/main/sphinx/sql/drop-materialized-view.rst [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+the materialized view does not exist.
+
+Examples
+--------
+
+Drop the materialized view ``orders_by_date``::
+
+    DROP MATERIALIZED VIEW orders_by_date;
+
+Drop the view ``orders_by_date`` if it exists::
```
Comment (anjalinorwood) [inline docs/src/main/sphinx/sql/drop-materialized-view.rst:27]: Drop the 'materialized' view ...

File: docs/src/main/sphinx/sql/refresh-materialized-view.rst (hunk 13 above)
Comment (anjalinorwood) [reply docs/src/main/sphinx/sql/refresh-materialized-view.rst:21]: The Trino Iceberg implementation does NOT drop the table. It deletes the data from the storage table and inserts the data that is the result of executing the materialized view query. 
Ref: https://github.com/trinodb/trino/blob/master/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java#L979

**Review by anjalinorwood — CHANGES_REQUESTED** [review]

File: docs/src/main/sphinx/sql/refresh-materialized-view.rst (hunk 13 above)
Comment (raunaqmorarka) [reply docs/src/main/sphinx/sql/refresh-materialized-view.rst:21]: Thanks for clarifying, i misunderstood the meaning of `executeDelete`

File: docs/src/main/sphinx/sql/create-materialized-view.rst [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+
+The optional ``COMMENT`` clause causes a ``string`` comment to be stored with
+the metadata about the materialized view. It can be displayed with the
+:doc:`show-create-materialized-view` statement and is available in
+``system.metadata.materialized_view_properties``.
+
+The optional ``WITH`` clause is used define properties to use for the
+materialized view creation. Separate multiple property/value pairs by commas.
+The supported properties are different for each connector and detailed in the
+SQL support section of the specific connector documentation.
```
Comment (m57lyra) [inline docs/src/main/sphinx/sql/create-materialized-view.rst:44]: connector's

File: docs/src/main/sphinx/sql/refresh-materialized-view.rst [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+--------
+
+.. code-block:: text
+
+    REFRESH MATERIALIZED VIEW view_name
+
+Description
+-----------
+
+Refresh the data stored in the materialized view ``view_name``. Data is
```
Comment (m57lyra) [inline docs/src/main/sphinx/sql/refresh-materialized-view.rst:15]: Refreshes

File: docs/src/main/sphinx/sql/drop-materialized-view.rst [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+--------
+
+.. code-block:: text
+
+    DROP MATERIALIZED VIEW [ IF EXISTS ] view_name;
+
+Description
+-----------
+
+Drop an existing materialized view.
```
Comment (m57lyra) [inline docs/src/main/sphinx/sql/drop-materialized-view.rst:15]: Drops

File: docs/src/main/sphinx/sql/create-materialized-view.rst (hunk 3 above)
Comment (m57lyra) [reply docs/src/main/sphinx/sql/create-materialized-view.rst:19]: Creates

File: docs/src/main/sphinx/sql/show-create-materialized-view.rst [hunk 20]
```diff
@@ -0,0 +1,22 @@
...
+--------
+
+.. code-block:: text
+
+    SHOW CREATE MATERIALIZED VIEW view_name;
+
+Description
+-----------
+
+Show the SQL statement that creates the specified materialized view.
```
Comment (m57lyra) [inline docs/src/main/sphinx/sql/show-create-materialized-view.rst:15]: Shows

File: docs/src/main/sphinx/sql/create-materialized-view.rst (hunk 14 above)
Comment (mosabua) [reply docs/src/main/sphinx/sql/create-materialized-view.rst:62]: updated

File: docs/src/main/sphinx/sql/refresh-materialized-view.rst (hunk 18 above)
Comment (mosabua) [reply docs/src/main/sphinx/sql/refresh-materialized-view.rst:15]: Thats against the style of all the other language docs. Leaving as is.

File: docs/src/main/sphinx/sql/create-materialized-view.rst (hunk 17 above)
Comment (mosabua) [reply docs/src/main/sphinx/sql/create-materialized-view.rst:44]: done

File: docs/src/main/sphinx/sql/create-materialized-view.rst [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+    [ COMMENT string ]
+    [ WITH properties ]
+    AS query
+
+Description
+-----------
+
+Create a new materialized view of a :doc:`select` query. The view is a physical
+manifestation of the query results at time of creation. The data is stored and
+can be referenced by future queries. Since the data is stored directly, these
```
[the remark this answers is not in the document]
Comment (mosabua) [reply docs/src/main/sphinx/sql/create-materialized-view.rst:21]: reworded the whole section

File: docs/src/main/sphinx/sql/create-materialized-view.rst (hunk 3 above)
Comment (mosabua) [reply docs/src/main/sphinx/sql/create-materialized-view.rst:19]: done with some more rewording

File: docs/src/main/sphinx/sql/drop-materialized-view.rst (hunk 12 above)
Comment (mosabua) [reply docs/src/main/sphinx/sql/drop-materialized-view.rst:17]: Not sure what the message is and hard to keep in sync. Leaving as is for now as this is a V1 doc set only

File: docs/src/main/sphinx/sql/refresh-materialized-view.rst (hunk 13 above)
Comment (mosabua) [reply docs/src/main/sphinx/sql/refresh-materialized-view.rst:21]: I think we can leave this info out actually .. removing for now. We can add more details in the specific connector docs or here at a later stage.

File: docs/src/main/sphinx/sql/create-materialized-view.rst (hunk 10 above)
Comment (mosabua) [reply docs/src/main/sphinx/sql/create-materialized-view.rst:57]: this is existing style in all the other language docs so I am staying consistent for now .. but I agree with your preference.

File: docs/src/main/sphinx/sql/create-materialized-view.rst (hunk 9 above)
Comment (mosabua) [reply docs/src/main/sphinx/sql/create-materialized-view.rst]: same style as all other docs ..leaving as is

File: docs/src/main/sphinx/sql/create-materialized-view.rst (hunk 8 above)
Comment (mosabua) [reply docs/src/main/sphinx/sql/create-materialized-view.rst:44]: Done

File: docs/src/main/sphinx/sql/create-materialized-view.rst (hunk 7 above)
Comment (mosabua) [reply docs/src/main/sphinx/sql/create-materialized-view.rst:39]: updated

File: docs/src/main/sphinx/sql/create-materialized-view.rst (hunk 6 above)
Comment (mosabua) [reply docs/src/main/sphinx/sql/create-materialized-view.rst:41]: reworded and corrected in more detail

File: docs/src/main/sphinx/sql/create-materialized-view.rst (hunk 5 above)
Comment (mosabua) [reply docs/src/main/sphinx/sql/create-materialized-view.rst]: done

File: docs/src/main/sphinx/sql/create-materialized-view.rst (hunk 4 above)
Comment (mosabua) [reply docs/src/main/sphinx/sql/create-materialized-view.rst]: done

File: docs/src/main/sphinx/sql/create-materialized-view.rst (hunk 2 above)
Comment (mosabua) [reply docs/src/main/sphinx/sql/create-materialized-view.rst]: I am leaving this out for now and have reworded that section a bit as well

File: docs/src/main/sphinx/sql/create-materialized-view.rst (hunk 1 above)
Comment (mosabua) [reply docs/src/main/sphinx/sql/create-materialized-view.rst:69]: all updated along that info

**Review by jhlodin — APPROVED** [review]

**Review by rosewms — APPROVED** [review]

File: docs/src/main/sphinx/sql/drop-materialized-view.rst [hunk 22]
```diff
@@ -0,0 +1,36 @@
+======================
+DROP MATERIALIZED VIEW
+======================
+
+Synopsis
+--------
+
+.. code-block:: text
+
+    DROP MATERIALIZED VIEW [ IF EXISTS ] view_name;
```
Comment (electrum) [inline docs/src/main/sphinx/sql/drop-materialized-view.rst:10]: Remove semicolon

File: docs/src/main/sphinx/sql/show-create-materialized-view.rst [hunk 23]
```diff
@@ -0,0 +1,22 @@
+=============================
+SHOW CREATE MATERIALIZED VIEW
+=============================
+
+Synopsis
+--------
+
+.. code-block:: text
+
+    SHOW CREATE MATERIALIZED VIEW view_name;
```
Comment (electrum) [inline docs/src/main/sphinx/sql/show-create-materialized-view.rst:10]: Remove semicolon

File: docs/src/main/sphinx/sql/create-materialized-view.rst [hunk 24]
```diff
@@ -0,0 +1,101 @@
...
+        SELECT orderdate, sum(totalprice) AS price
+        FROM orders
+        GROUP BY orderdate
+
+Create a materialized view with a comment and partitioning on two fields in the
+storage::
+
+    CREATE MATERIALIZED VIEW orders_nation_mkgsegment
+    COMMENT 'Orders with nation and market segment data'
+    WITH ( partitioning = ARRAY['mktsegment','nationkey'] )
```
Comment (electrum) [inline docs/src/main/sphinx/sql/create-materialized-view.rst:81]: Space after comma

File: docs/src/main/sphinx/sql/create-materialized-view.rst [hunk 25]
```diff
@@ -0,0 +1,101 @@
...
+Create a simple materialized view ``cancelled_orders`` over the ``orders`` table
+that only includes cancelled orders. Note that ``orderstatus`` is a numeric
+value that is potentially meaningless to a consumer, yet the name of the view
+clarifies the content::
+
+    CREATE MATERIALIZED VIEW cancelled_orders
+    AS
+        SELECT orderkey, totalprice
+        FROM orders
+        WHERE orderstatus = 3
```
Comment (electrum) [inline docs/src/main/sphinx/sql/create-materialized-view.rst:65]: Add semicolons after these since they are full queries. This will differentiate from the `WITH` fragment below

File: docs/src/main/sphinx/sql/refresh-materialized-view.rst [hunk 26]
```diff
@@ -0,0 +1,35 @@
...
+``view_name``. The materialized view must be defined with
+:doc:`create-materialized-view`. Data is retrieved from the underlying tables
+accessed by the defined query.
+
+The initial population of the materialized view is typically processing
+intensive since it reads the data from the source tables and performs physical
+write operations.
+
+The refresh operation can be less intensive, if the underlying data has not
+changed and the connector has implemented mechanism to be aware of that.
```
Comment (electrum) [inline docs/src/main/sphinx/sql/refresh-materialized-view.rst:25]: "a mechanism"

File: docs/src/main/sphinx/sql/refresh-materialized-view.rst [hunk 27]
```diff
@@ -0,0 +1,35 @@
...
+Connectors can also simple remove the data and reinsert it, or even perform
```
Comment (electrum) [inline docs/src/main/sphinx/sql/refresh-materialized-view.rst:26]: Let's remove this sentence. I tried to rewrite it, but I'm not sure it's needed

File: docs/src/main/sphinx/sql/refresh-materialized-view.rst [hunk 28]
```diff
@@ -0,0 +1,35 @@
...
+update and insert operations as necessary. The specific implementation varies by
```
Comment (electrum) [inline docs/src/main/sphinx/sql/refresh-materialized-view.rst:27]: maybe expand to

> implementation and performance

**Review by electrum — APPROVED** [review]

File: docs/src/main/sphinx/sql/drop-materialized-view.rst (hunk 22 above)
Comment (mosabua) [reply docs/src/main/sphinx/sql/drop-materialized-view.rst:10]: done

File: docs/src/main/sphinx/sql/create-materialized-view.rst (hunk 24 above)
Comment (mosabua) [reply docs/src/main/sphinx/sql/create-materialized-view.rst:81]: done

File: docs/src/main/sphinx/sql/create-materialized-view.rst (hunk 25 above)
Comment (mosabua) [reply docs/src/main/sphinx/sql/create-materialized-view.rst:65]: done

File: docs/src/main/sphinx/sql/refresh-materialized-view.rst (hunk 27 above)
Comment (mosabua) [reply docs/src/main/sphinx/sql/refresh-materialized-view.rst:26]: done

File: docs/src/main/sphinx/sql/refresh-materialized-view.rst (hunk 26 above)
Comment (mosabua) [reply docs/src/main/sphinx/sql/refresh-materialized-view.rst:25]: done

File: docs/src/main/sphinx/sql/refresh-materialized-view.rst (hunk 28 above)
Comment (mosabua) [reply docs/src/main/sphinx/sql/refresh-materialized-view.rst:27]: done

File: docs/src/main/sphinx/sql/show-create-materialized-view.rst (hunk 23 above)
Comment (mosabua) [reply docs/src/main/sphinx/sql/show-create-materialized-view.rst:10]: done

**Comment by mosabua** [comment]
Thanks for the help everyone and merge @electrum !

**Comment by tooptoop4** [comment]
where is the data in the MV stored? @mosabua

**Comment by bitsondatadev** [comment]
@tooptoop4, they are stored in the object store. @anjalinorwood  talks more about it on this episode and I created some sequence diagrams to relay what's going on in different views here: https://trino.io/episodes/18.html

**Comment by anjalinorwood** [comment]
> where is the data in the MV stored? @mosabua 

Each materialized view is in reality a view + a storage iceberg table. The storage table name is stored as a materialized view property. The data is stored in that storage table.
